spec/requests/halls_spec.rb

```diff
@@ -0,0 +1,345 @@
+require 'rails_helper'
+
+RSpec.describe "/halls", type: :request do
+  let(:user) { create :user }
+  let(:manager) { create :user, email: "testmanager@test.com", role: :manager }
+  describe "GET /halls" do
+    context "when no user" do
+      it "redirects to sign_in" do
+        get("/halls")
+        expect(response).to redirect_to("/users/sign_in")
+      end
+
+      it "returns redirect status" do
+        get("/halls")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user without permission" do
+      before { sign_in user }
+      it "redirects to root" do
+        get("/halls")
+        expect(response).to redirect_to("/")
+      end
+
+      it "returns redirect status" do
+        get("/halls")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user with permission" do
+      before { sign_in manager }
+      it "returns successful response" do
+        get("/halls")
+        expect(response.status).to eq(200)
+      end
+
+      it "renders proper template" do
+        get("/halls")
+        expect(response).to render_template("halls/index")
+      end
+    end
+  end
+
+  describe "GET /halls/new" do
+    context "when no user" do
+      it "redirects to sign_in" do
+        get("/halls/new")
+        expect(response).to redirect_to("/users/sign_in")
+      end
+
+      it "returns redirect status" do
+        get("/halls/new")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user without permission" do
+      before { sign_in user }
+      it "redirects to root" do
+        get("/halls/new")
+        expect(response).to redirect_to("/")
+      end
+
+      it "returns redirect status" do
+        get("/halls/new")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user with permission" do
+      before { sign_in manager }
+      it "returns successful response" do
+        get("/halls/new")
+        expect(response.status).to eq(200)
+      end
+
+      it "renders proper template" do
+        get("/halls/new")
+        expect(response).to render_template("halls/new")
+      end
+    end
+  end
+
+  describe "POST /halls" do
+    let(:params) do
+      {
+        hall: {
+          name:,
+          capacity:
+        }
+      }
+    end
+
+    let(:name) { "Sala #{Faker::Number.number(digits: 1)}" }
+    let(:capacity) { Faker::Number.number(digits: 2) }
```

Comment:
Use FactoryBot's `attributes_for` to generate param hashes
```suggestion
    let(:params) { attributes_for(:hall, name:, capacity:) }
    let(:name) { "Sala #{Faker::Number.number(digits: 1)}" }
    let(:capacity) { Faker::Number.number(digits: 2) }
```

Reply:
had to put it like this: 
  ```
let(:params) do
        {
          hall: attributes_for(:hall, name:, capacity:)
        }
      end
```
otherwise controller returned the parameters error.
```
Failure/Error: params.require(:hall).permit(:capacity, :name)
     
     ActionController::ParameterMissing:
       param is missing or the value is empty: hall
```